File: src/main/java/io/github/dylmeadows/eontimer/core/timer/CustomTimerFactory.java
package io.github.dylmeadows.eontimer.core.timer;

import io.github.dylmeadows.eontimer.model.Stage;
import io.github.dylmeadows.eontimer.model.timer.CustomTimerModel;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class CustomTimerFactory implements TimerFactory {

    private final CustomTimerModel timerModel;

    @Override
    public List<Stage> createTimer() {
        return getStages();
    }

    private List<Stage> getStages() {
        return timerModel.getStages()
            .stream()
            .map(Stage::new)
            .collect(Collectors.toList());
    }
}

File: src/main/kotlin/io/github/dylmeadows/eontimer/service/factory/CustomTimerFactory.kt
package io.github.dylmeadows.eontimer.service.factory

import io.github.dylmeadows.eontimer.model.TimerState
import io.github.dylmeadows.eontimer.model.timer.CustomTimerModel
import io.github.dylmeadows.eontimer.util.asFlux
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.time.Duration
import javax.annotation.PostConstruct

@Component
class CustomTimerFactory @Autowired constructor(
    private val timerState: TimerState,
    private val customTimerModel: CustomTimerModel) : TimerFactory {

    override val stages: List<Duration>
        get() = customTimerModel.stages
            .map { Duration.ofMillis(it.length) }

    @PostConstruct
    private fun initialize() {
        customTimerModel.stages.asFlux()
            .subscribe {
                timerState.update(stages)
            }
    }

    override fun calibrate() = Unit
}